std/internal/cstring.d

```diff
@@ -132,55 +132,69 @@ auto tempCString(To = char, From)(From str)
         }
 
         To[256 / To.sizeof] _buff;  // the 'small string optimization'
+
+        static Res trustedVoidInit() { Res res = void; return res; }
     }
 
-    Res res = void;     // expensive to fill _buff[]
+    Res res = Res.trustedVoidInit();     // expensive to fill _buff[]
 
     // Note: res._ptr can't point to res._buff as structs are movable.
 
-    import std.utf : byUTF;
-
-    To* p      = res._buff.ptr;
-    size_t len = res.buffLength;
+    To[] p = res._buff[0 .. Res.buffLength];
     size_t i;
 
+    static To[] trustedRealloc(To[] buf, size_t i, To* resptr, size_t strLength)
+        @trusted @nogc nothrow
+    {
+        pragma(inline, false);  // because it's rarely called
+
+        import core.exception   : onOutOfMemoryError;
+        import core.stdc.string : memcpy;
+        import core.stdc.stdlib : malloc, realloc;
+
+        auto p = buf.ptr;
+        auto len = buf.length;
+        if (len >= size_t.max / (2 * To.sizeof))
+            onOutOfMemoryError();
+        size_t newlen = len * 3 / 2;
+        if (p == resptr)
+        {
+            if (newlen <= strLength)
+                newlen = strLength + 1; // +1 for terminating 0
+            p = cast(To*)malloc(newlen * To.sizeof);
+            if (!p)
+                onOutOfMemoryError();
+            memcpy(p, resptr, i * To.sizeof);
+        }
+        else
+        {
+            p = cast(To*)realloc(p, newlen * To.sizeof);
+            if (!p)
+                onOutOfMemoryError();
+        }
+        return p[0 .. newlen];
+    }
+
+    size_t strLength;
+    static if (hasLength!From)
+    {
+        strLength = str.length;
+    }
+    import std.utf : byUTF;
     static if (isSomeString!From)
-        auto r = cast(const(CF)[])str;
+        auto r = cast(const(CF)[])str;  // because inout(CF) causes problems with byUTF
     else
         alias r = str;
     foreach (const c; byUTF!(Unqual!To)(r))
     {
-        if (i + 1 == len)
+        if (i + 1 == p.length)
```

Comment:
Can you use different variable names for your nested function vs. the buffer? I was confused here thinking `p` was a pointer from looking at line 155.